Core/clim-basic/extended-streams/recording.lisp

```diff
@@ -143,44 +143,47 @@ all of FUNCTION-ARGS as APPLY arguments."
+                                   spec
+                                   (list (alexandria:make-keyword spec)
+                                         spec))
+          for suppliedp = (gensym (format nil "~A-P" keyword))
+          for slotp = (if (and (consp spec) (>= (length spec) 3))
+                          (third spec)
+                          t)
+          when slotp do (push name class-slots)
+          do (push `((,keyword ,name) nil ,suppliedp) key-args)
+          do (push (cons name suppliedp) key-arg-alist))
     `(progn
        (defmethod output-record-equal and ((record ,record-type)
                                            (record2 ,record-type))
-         (macrolet ((if-supplied ((var &optional (type t)) &body supplied-body)
-                      (declare (ignore var type))
-                      `(progn ,@supplied-body)))
-           (with-slots ,slot-names record2
+         (macrolet ((if-supplied ((var &optional (type t))
+                                  slot-form &optional arg-form)
+                      (declare (ignore type arg-form))
+                      (when (find var ',class-slots :test #'eq)
+                        slot-form)))
+           (with-slots ,(nreverse class-slots) record2
              ,@body)))
        (defmethod match-output-records-1 and ((record ,record-type)
-                                              &key ,@key-args)
-         (macrolet ((if-supplied ((var &optional (type t)) &body supplied-body)
-                      (let ((supplied-var (cdr (assoc var ',key-arg-alist))))
-                        (unless supplied-var
-                          (error "Unknown slot ~S" var))
-                        `(or (null ,supplied-var)
-                             ,@(if (eq type t)
-                                   `((progn ,@supplied-body))
-                                   `((if (typep ,var ',type)
-                                         (progn ,@supplied-body)
-                                         (error 'type-error
-                                                :datum ,var
-                                                :expected-type ',type))))))))
+                                              &key ,@(nreverse key-args))
+         (macrolet ((if-supplied ((var &optional (type t))
+                                  slot-form &optional (arg-form slot-form))
+                      (let ((supplied-var (or (cdr (assoc var ',key-arg-alist))
+                                              (error "Unknown slot ~S" var))))
+                        `(when ,supplied-var
+                           ,(if (eq type t)
+                                arg-form
+                                `(if (typep ,var ',type)
+                                     ,arg-form
+                                     (error 'type-error
+                                            :datum ,var
+                                            :expected-type ',type)))))))
```

Comment:
I was actually wondering why check the type at all? I seems completely arbitrary.

Reply:
when the type is known I think that it makes sense to check it. We should do it in more places though. Example:

(if-supplied (ink design)
    (design-equalp (slot-value record 'ink) ink))

if ink is not a design, then we will never call design-equalp (in code the type is not declared though). It seems to be mostly checked for coordinates – in this case that may improve the performance, because we know that we can inline some operations.

Reply:
I would prefer to just fix the regression for now. We can work on the type checking later.

Reply:
I'm not asking you to add annotations to other if-supplied forms, only to change the expansion (and I doubt we will get back to it before 2025). It doesn't matter though, because the change would be cosmetic.

That said, I've spotted another problem here: you change the semantics of the `if-supplied` macro.

Previously: when the argument is not supplied, returns T
Currently: when the argument is not supplied returns NIL

The latter is incorrect, consider
```
(defrecord-predicate gs-ink-mixin (ink)
  (if-supplied (ink)
    (design-equalp (slot-value record 'ink) ink)))
```
when the ink is not supplied, then predicate should default to T. Please bring back the old form (especially if we agree about not changing the lambda list of the macro -- this change won't be necessary at all then.